File: HaNoiTower/hanoitower.cpp
#include <bits/stdc++.h>

using namespace std;

void guide(int d, int s, int t)
{
    cout << "Move disk " << d << " from rod " << s << " to rod " << t << endl;
}

void HaNoiTower(int disk, int source, int target)
{
    int spare; // spare rod

    if (disk == 1)
    {
        guide(disk, source, target);
        return;
    }

    spare = 6 - (source + target);
    
    // move upper (n - 1) disks to the spare rod
    HaNoiTower(disk - 1, source, spare);

    // move the n-th disk (the bottom disk) to the target rod
    guide(disk, source, target);

    // move (n - 1) remaining disks to the target rod
    HaNoiTower(disk - 1, spare, target);
}

int main()
{
    int n = 4; // the number of disks
    int s = 1; // source rod
    int t = 3; // target rod
    HaNoiTower(n, s, t);

    return 0;
}

File: HaNoiTower/hanoitower.py
def guide(d, s, t):
    print(f'Move disc {d} from rod {s} to rod {t}')


def hanoi_tower(disc, source, target):
    spare = 0

    if disc == 1:
        guide(disc, source, target)
        return
    
    spare = 6 - (source + target)

    # Di chuyển n - 1 đĩa nằm trên sang cột trung gian
    hanoi_tower(disc - 1, source, spare)

    # Di chuyển đĩa thứ n (nằm dưới cùng) sang cột đích 
    guide(disc, source, target)

    # Di chuyển n - 1 đĩa còn lại từ cột trung gian sang cột đích
    hanoi_tower(disc - 1, spare, target)


if __name__ == '__main__':
    n = 4 # số đĩa
    s = 1 # cột nguồn
    t = 3 # cột đích

    hanoi_tower(n, s, t)

    